File: fifteen-exercises/rotate-image.rs
impl Solution {
    pub fn rotate(matrix: &mut Vec<Vec<i32>>) {
        fn swap(matrix: &mut Vec<Vec<i32>>, i1: usize, j1: usize, i2: usize, j2: usize) {
            let temp = matrix[i1][j1];
            matrix[i1][j1] = matrix[i2][j2];
            matrix[i2][j2] = temp;
        }
        let n = matrix.len();
        for i in 0..n {
            for j in 0..i + 1 {
                swap(matrix, i, j, j, i);
            }
        }
        for i in 0..n {
            for j in 0..n / 2 {
                swap(matrix, i, j, i, n - 1 - j);
            }
        }
    }
}

struct Solution;

fn main() {
    let m = &mut vec![vec![1, 2, 3], vec![4, 5, 6], vec![7, 8, 9]];
    Solution::rotate(m);
    assert_eq!(m, &mut vec![vec![7, 4, 1], vec![8, 5, 2], vec![9, 6, 3]]);
    println!("ok")
}

File: fifteen-exercises/rotate-image.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <functional>

using namespace std;

class Solution {
public:
    void rotate(vector<vector<int>> &matrix) {
        std::size_t n = matrix.size();
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < i + 1; ++j) {
                swap(matrix[i][j], matrix[j][i]);
            }
        }
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n / 2; ++j) {
                swap(matrix[i][j], matrix[i][n - 1 - j]);
            }
        }
    }
};

int main() {
    cout << "ok" << endl;
    return 0;
}
